src/middlewared/middlewared/plugins/account_/privilege.py

```diff
@@ -0,0 +1,161 @@
+import enum
+import errno
+
+from middlewared.schema import accepts, Bool, Dict, Int, List, Ref, Str, Patch
+from middlewared.service import CallError, CRUDService, private, ValidationErrors
+import middlewared.sqlalchemy as sa
+
+
+class BuiltinPrivileges(enum.Enum):
+    LOCAL_ADMINISTRATOR = "LOCAL_ADMINISTRATOR"
+
+
+class PrivilegeModel(sa.Model):
+    __tablename__ = "account_privilege"
+
+    id = sa.Column(sa.Integer(), primary_key=True)
+    builtin_name = sa.Column(sa.String(200), nullable=True)
+    name = sa.Column(sa.String(200))
+    local_groups = sa.Column(sa.JSON(type=list))
+    ds_groups = sa.Column(sa.JSON(type=list))
+    allowlist = sa.Column(sa.JSON(type=list))
+    web_shell = sa.Column(sa.Boolean())
+
+
+class PrivilegeService(CRUDService):
+
+    keys = {}
+
+    class Config:
+        namespace = "privilege"
+        datastore = "account.privilege"
+        datastore_extend = "privilege.item_extend"
+        datastore_extend_context = "privilege.item_extend_context"
+        cli_namespace = "auth.privilege"
+
+    ENTRY = Dict(
+        "privilege_entry",
+        Int("id"),
+        Str("builtin_name", null=True),
+        Str("name", required=True, empty=False),
+        List("local_groups", items=[Int("local_group")]),
+        List("ds_groups", items=[Int("ds_group")]),
+        List("allowlist", items=[Ref("allowlist_item")]),
+        Bool("web_shell", required=True),
+    )
+
+    @private
+    async def item_extend_context(self, rows, extra):
+        return {
+            "local_groups": {
+                g["id"]: {"id": g["id"], "name": g["group"]}
+                for g in await self.middleware.call("datastore.query", "account.bsdgroups", [], {"prefix": "bsdgrp_"})
+            },
+        }
+
+    @private
+    async def item_extend(self, item, context):
+        item["local_groups"] = [context["local_groups"][group_id]
+                                for group_id in item["local_groups"]
+                                if group_id in context["local_groups"]]
+        return item
+
+    @accepts(Patch(
+        "privilege_entry",
+        "privilege_create",
+        ("rm", {"name": "builtin_name"}),
+    ))
+    async def do_create(self, data):
+        """
+        Creates a privilege.
+
+        `local_groups` is a list of local user account groups that gain this privilege.
+
+        `ds_groups` is list of Directory Service group IDs that will gain this privilege.
+        """
```

Comment:
Can't users specify `allowlist` `name` and `web shell`? Maybe expand docstring.

Reply:
Done